sound/soc/sof/intel/hda.c

```diff
@@ -251,15 +261,56 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 	}
 
 	/* codec detection */
-	if (!bus->codec_mask)
+	if (!bus->codec_mask) {
 		dev_info(bus->dev, "no hda codecs found!\n");
-	else
-		dev_info(bus->dev, "hda codecs found, mask %lx!\n", bus->codec_mask);
+	} else {
+		dev_info(bus->dev, "hda codecs found, mask %lx\n",
+			 bus->codec_mask);
+
+		for (i = 0; i < HDA_MAX_CODECS; i++) {
+			if (bus->codec_mask & (1 << i))
+				codec_num++;
+		}
+
+		/*
+		 * If no machine driver is found, then:
+		 *
+		 * hda machine driver is used if :
+		 * 1. there are HDMI codec and other HDA codecs
+		 * 2. only HDMI codec and NOCODEC is not enabled
+		 *
+		 * nocodec machine driver would be used if:
+		 * 1. only HDMI codec and but NOCODEC is enabled
+		 * 2. there is no codec found
+		 */
+		if (!pdata->machine) {
+			if (codec_num > 1)
+				hda_mach = snd_soc_acpi_intel_hda_machines;
+#if !IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)
+			else if (HDMI_CODEC(bus->codec_mask))
+				hda_mach = snd_soc_acpi_intel_hda_machines;
+#endif
+
```

Comment:
Actually for the first case we need to double check HDMI is present (codec_num>1 is not enough in theory, there could be two HDaudio codecs which isn't supported). Something like this would work:

````c
if (HDA_IDISP_CODEC(bus->codec_mask)) {
    if (codec_num == 2 || !IS_ENABLED(CONFIG_SND_SOC_SOF_NOCODEC)) {
        hda_mach = snd_soc_acpi_intel_hda_machines;
    }
}
````

Reply:
yes, this I am uncertain about. Thanks!